File: dart_sample/lib/demos/general/proxy_widgets_demo_page.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';

import '../../counter_widgets.dart';

class ProxyWidgetsDemoPage extends StatefulWidget {
  const ProxyWidgetsDemoPage({super.key});

  @override
  State<ProxyWidgetsDemoPage> createState() => _ProxyWidgetsDemoPageState();
}

class _ProxyWidgetsDemoPageState extends State<ProxyWidgetsDemoPage> {
  double _opacity = 0.8;
  double _shiftX = 0;
  bool _tightClip = true;
  double _fractionalShift = 0;
  int _quarterTurns = 0;

  @override
  Widget build(BuildContext context) {
    final Rect clip = _tightClip
        ? const Rect.fromLTWH(0, 0, 120, 80)
        : const Rect.fromLTWH(0, 0, 190, 110);

    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 10,
      children: <Widget>[
        const Text(
          'Proxy widgets: transforms + clips',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'Use controls to fade a high-contrast black card over white canvas.',
          style: TextStyle(fontSize: 14, color: Colors.black54),
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'Opacity -',
              onTap: () => _changeOpacity(-0.3),
              width: 96,
              background: const Color(0xFFDCE3ED),
            ),
            _buildButton(
              label: 'Opacity +',
              onTap: () => _changeOpacity(0.3),
              width: 96,
              background: const Color(0xFFDCE3ED),
            ),
            _buildButton(
              label: 'Reset',
              onTap: _reset,
              width: 88,
              background: const Color(0xFFE9F5EC),
            ),
          ],
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'Left',
              onTap: () => _move(-20),
              width: 88,
              background: const Color(0xFFF3E8D8),
            ),
            _buildButton(
              label: 'Right',
              onTap: () => _move(20),
              width: 88,
              background: const Color(0xFFF3E8D8),
            ),
            _buildButton(
              label: _tightClip ? 'Clip: tight' : 'Clip: wide',
              onTap: _toggleClip,
              width: 104,
              background: const Color(0xFFE8EDF9),
            ),
          ],
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'Opacity 0',
              onTap: () => _setOpacity(0),
              width: 96,
              background: const Color(0xFFF6E0E0),
            ),
            _buildButton(
              label: 'Opacity 1',
              onTap: () => _setOpacity(1),
              width: 96,
              background: const Color(0xFFE0F0E7),
            ),
          ],
        ),
        Text(
          'opacity=${_opacity.toStringAsFixed(2)}, shiftX=${_shiftX.toStringAsFixed(0)}, clip=${_tightClip ? 'tight' : 'wide'}',
          style: const TextStyle(fontSize: 12, color: Colors.blueGrey),
        ),
        Container(
          width: 220,
          height: 140,
          color: const Color(0xFFE7EDF6),
          padding: const EdgeInsets.all(8),
          child: ClipRect(
            clipper: _FixedRectClipper(clip),
            child: Transform.translate(
              offset: Offset(_shiftX, 10),
              child: Opacity(
                opacity: _opacity,
                child: Container(
                  width: 140,
                  height: 90,
                  color: const Color(0xFF111111),
                  padding: const EdgeInsets.all(8),
                  child: const Text(
                    'Layer',
                    style: TextStyle(fontSize: 14, color: Colors.white),
                  ),
                ),
              ),
            ),
          ),
        ),
        const Text(
          'FractionalTranslation + RotatedBox',
          style: TextStyle(fontSize: 14, color: Colors.black),
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'Shift',
              onTap: _cycleFractionalShift,
              width: 88,
              background: const Color(0xFFE8EDF9),
            ),
            _buildButton(
              label: 'Rotate',
              onTap: _rotateQuarterTurn,
              width: 88,
              background: const Color(0xFFF3E8D8),
            ),
            Text(
              'fraction=${_fractionalShift.toStringAsFixed(1)}, turns=$_quarterTurns',
              style: const TextStyle(fontSize: 12, color: Colors.blueGrey),
            ),
          ],
        ),
        Row(
          spacing: 16,
          children: <Widget>[
            Container(
              width: 120,
              height: 80,
              color: const Color(0xFFE7EDF6),
              child: Center(
                child: FractionalTranslation(
                  translation: Offset(_fractionalShift, 0),
                  child: Container(
                    width: 56,
                    height: 32,
                    color: const Color(0xFF6750A4),
                    child: const Center(
                      child: Text(
                        'Shift',
                        style: TextStyle(fontSize: 12, color: Colors.white),
                      ),
                    ),
                  ),
                ),
              ),
            ),
            Container(
              width: 120,
              height: 80,
              color: const Color(0xFFE7EDF6),
              child: Center(
                child: RotatedBox(
                  quarterTurns: _quarterTurns,
                  child: Container(
                    width: 64,
                    height: 28,
                    color: const Color(0xFF386A20),
                    child: const Center(
                      child: Text(
                        'Rotate',
                        style: TextStyle(fontSize: 12, color: Colors.white),
                      ),
                    ),
                  ),
                ),
              ),
            ),
          ],
        ),
        const Text(
          'ClipOval + ClipPath',
          style: TextStyle(fontSize: 14, color: Colors.black),
        ),
        Row(
          spacing: 16,
          children: <Widget>[
            SizedBox(
              width: 96,
              height: 72,
              child: ClipOval(
                child: ColoredBox(
                  color: Color(0xFF6750A4),
                  child: Center(
                    child: Text(
                      'Oval',
                      style: TextStyle(fontSize: 13, color: Colors.white),
                    ),
                  ),
                ),
              ),
            ),
            SizedBox(
              width: 96,
              height: 72,
              child: ClipPath(
                clipper: _TrianglePathClipper(),
                child: ColoredBox(
                  color: Color(0xFF386A20),
                  child: Center(
                    child: Text(
                      'Path',
                      style: TextStyle(fontSize: 13, color: Colors.white),
                    ),
                  ),
                ),
              ),
            ),
          ],
        ),
      ],
    );
  }

  Widget _buildButton({
    required String label,
    required VoidCallback onTap,
    required double width,
    required Color background,
  }) {
    return SizedBox(
      width: width,
      child: CounterTapButton(
        label: label,
        onTap: onTap,
        background: background,
        foreground: Colors.black,
        fontSize: 12,
        padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
      ),
    );
  }

  void _changeOpacity(double delta) {
    setState(() {
      _opacity = (_opacity + delta).clamp(0.0, 1.0);
    });
  }

  void _setOpacity(double value) {
    setState(() {
      _opacity = value.clamp(0.0, 1.0);
    });
  }

  void _move(double delta) {
    setState(() {
      _shiftX = math.max(-40, math.min(80, _shiftX + delta));
    });
  }

  void _toggleClip() {
    setState(() {
      _tightClip = !_tightClip;
    });
  }

  void _cycleFractionalShift() {
    setState(() {
      _fractionalShift = _fractionalShift >= 0.5
          ? -0.5
          : _fractionalShift + 0.5;
    });
  }

  void _rotateQuarterTurn() {
    setState(() {
      _quarterTurns = (_quarterTurns + 1) % 4;
    });
  }

  void _reset() {
    setState(() {
      _opacity = 0.8;
      _shiftX = 0;
      _tightClip = true;
      _fractionalShift = 0;
      _quarterTurns = 0;
    });
  }
}

class _FixedRectClipper extends CustomClipper<Rect> {
  const _FixedRectClipper(this.rect);

  final Rect rect;

  @override
  Rect getClip(Size size) => rect;

  @override
  bool shouldReclip(_FixedRectClipper oldClipper) => oldClipper.rect != rect;
}

class _TrianglePathClipper extends CustomClipper<Path> {
  const _TrianglePathClipper();

  @override
  Path getClip(Size size) {
    return Path()
      ..moveTo(size.width / 2, 0)
      ..lineTo(size.width, size.height)
      ..lineTo(0, size.height)
      ..close();
  }

  @override
  bool shouldReclip(_TrianglePathClipper oldClipper) => false;
}

File: src/Sample/Plumix.Sample/Demos/General/ProxyWidgetsDemoPage.cs
using System;
using Avalonia;
using Avalonia.Media;
using Plumix.Rendering;
using Plumix.Widgets;
using Path = Plumix.UI.Path;

// Dart parity source (reference): dart_sample/lib/proxy_widgets_demo_page.dart (exact sample parity)

namespace Plumix;

public sealed class ProxyWidgetsDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new ProxyWidgetsDemoPageState();
    }
}

internal sealed class ProxyWidgetsDemoPageState : State
{
    private double _opacity = 0.8;
    private double _shiftX;
    private bool _tightClip = true;
    private double _fractionalShift;
    private int _quarterTurns;

    public override Widget Build(BuildContext context)
    {
        var clip = _tightClip
            ? new Rect(0, 0, 120, 80)
            : new Rect(0, 0, 190, 110);

        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 10,
            children:
            [
                new Text(
                    "Proxy widgets: transforms + clips",
                    fontSize: 20,
                    color: Colors.Black),
                new Text(
                    "Use controls to fade a high-contrast black card over white canvas.",
                    fontSize: 14,
                    color: Colors.DimGray),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("Opacity -", () => ChangeOpacity(-0.3), width: 96, colorHex: "#FFDCE3ED"),
                        BuildButton("Opacity +", () => ChangeOpacity(0.3), width: 96, colorHex: "#FFDCE3ED"),
                        BuildButton("Reset", Reset, width: 88, colorHex: "#FFE9F5EC"),
                    ]),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("Left", () => Move(-20), width: 88, colorHex: "#FFF3E8D8"),
                        BuildButton("Right", () => Move(20), width: 88, colorHex: "#FFF3E8D8"),
                        BuildButton(_tightClip ? "Clip: tight" : "Clip: wide", ToggleClip, width: 104, colorHex: "#FFE8EDF9"),
                    ]),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("Opacity 0", () => SetOpacity(0), width: 96, colorHex: "#FFF6E0E0"),
                        BuildButton("Opacity 1", () => SetOpacity(1), width: 96, colorHex: "#FFE0F0E7"),
                    ]),
                new Text(
                    $"opacity={_opacity:0.00}, shiftX={_shiftX:0}, clip={(_tightClip ? "tight" : "wide")}",
                    fontSize: 12,
                    color: Colors.DarkSlateGray),
                new Container(
                    width: 220,
                    height: 140,
                    color: Color.Parse("#FFE7EDF6"),
                    padding: new Thickness(8),
                    child: new ClipRect(
                        clipRect: clip,
                        child: new Plumix.Widgets.Transform(
                            transform: Matrix.CreateTranslation(_shiftX, 10),
                            child: new Opacity(
                                opacity: _opacity,
                                child: new Container(
                                    width: 140,
                                    height: 90,
                                    color: Color.Parse("#FF111111"),
                                    padding: new Thickness(8),
                                    child: new Text("Layer", fontSize: 14, color: Colors.White)))))),
                new Text("FractionalTranslation + RotatedBox", fontSize: 14, color: Colors.Black),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("Shift", CycleFractionalShift, width: 88, colorHex: "#FFE8EDF9"),
                        BuildButton("Rotate", RotateQuarterTurn, width: 88, colorHex: "#FFF3E8D8"),
                        new Text(
                            $"fraction={_fractionalShift:0.0}, turns={_quarterTurns}",
                            fontSize: 12,
                            color: Colors.DarkSlateGray),
                    ]),
                new Row(
                    spacing: 16,
                    children:
                    [
                        new Container(
                            width: 120,
                            height: 80,
                            color: Color.Parse("#FFE7EDF6"),
                            child: new Center(
                                new FractionalTranslation(
                                    translation: new Vector(_fractionalShift, 0),
                                    child: new Container(
                                        width: 56,
                                        height: 32,
                                        color: Color.Parse("#FF6750A4"),
                                        child: new Center(
                                            new Text("Shift", fontSize: 12, color: Colors.White)))))),
                        new Container(
                            width: 120,
                            height: 80,
                            color: Color.Parse("#FFE7EDF6"),
                            child: new Center(
                                new RotatedBox(
                                    quarterTurns: _quarterTurns,
                                    child: new Container(
                                        width: 64,
                                        height: 28,
                                        color: Color.Parse("#FF386A20"),
                                        child: new Center(
                                            new Text("Rotate", fontSize: 12, color: Colors.White)))))),
                    ]),
                new Text("ClipOval + ClipPath", fontSize: 14, color: Colors.Black),
                new Row(
                    spacing: 16,
                    children:
                    [
                        new SizedBox(
                            width: 96,
                            height: 72,
                            child: new ClipOval(
                                child: new ColoredBox(
                                    Color.Parse("#FF6750A4"),
                                    new Center(new Text("Oval", fontSize: 13, color: Colors.White))))),
                        new SizedBox(
                            width: 96,
                            height: 72,
                            child: new ClipPath(
                                clipper: new TrianglePathClipper(),
                                child: new ColoredBox(
                                    Color.Parse("#FF386A20"),
                                    new Center(new Text("Path", fontSize: 13, color: Colors.White))))),
                    ]),
            ]);
    }

    private Widget BuildButton(string label, Action onTap, double width, string colorHex)
    {
        return new SizedBox(
            width: width,
            child: new CounterTapButton(
                label: label,
                onTap: onTap,
                background: Color.Parse(colorHex),
                foreground: Colors.Black,
                fontSize: 12,
                padding: new Thickness(10, 8)));
    }

    private void ChangeOpacity(double delta)
    {
        SetState(() => _opacity = Math.Clamp(_opacity + delta, 0, 1));
    }

    private void SetOpacity(double value)
    {
        SetState(() => _opacity = Math.Clamp(value, 0, 1));
    }

    private void Move(double delta)
    {
        SetState(() => _shiftX = Math.Clamp(_shiftX + delta, -40, 80));
    }

    private void ToggleClip()
    {
        SetState(() => _tightClip = !_tightClip);
    }

    private void CycleFractionalShift()
    {
        SetState(() => _fractionalShift = _fractionalShift >= 0.5 ? -0.5 : _fractionalShift + 0.5);
    }

    private void RotateQuarterTurn()
    {
        SetState(() => _quarterTurns = (_quarterTurns + 1) % 4);
    }

    private void Reset()
    {
        SetState(() =>
        {
            _opacity = 0.8;
            _shiftX = 0;
            _tightClip = true;
            _fractionalShift = 0;
            _quarterTurns = 0;
        });
    }

    private sealed class TrianglePathClipper : CustomClipper<Path>
    {
        public override Path GetClip(Size size)
        {
            var path = new Path();
            path.MoveTo(size.Width / 2.0, 0);
            path.LineTo(size.Width, size.Height);
            path.LineTo(0, size.Height);
            path.Close();
            return path;
        }

        public override bool ShouldReclip(CustomClipper<Path> oldClipper) => false;
    }
}
